build/xwalk.gni

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2016 Intel Corporation. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# This file contains Crosswalk-feature-related build flags. Users should
+# import it into their build folder by run:
+#   echo 'import("//xwalk/build/xwalk.gni")' > out/your_folder/args.gn'
+# Then run:
+#   gn gen out/your_folder
+
+# Set it false when building content_shell.
+declare_args() {
```

Comment:
> Sorry I didn't get your points. My understanding: GN will always tries to check the BUILD.gn file in src/ directory.
> Only if it's not found, the build/secondary/xwalk/BUILD.gn will be checked. But in our case, the src/BUILD.gn is always there no matter you checkout "chromium-crosswalk only" or "chromium-crosswalk plus crosswalk." Could you please give more explanations on this?

Again, see GN's documentation: [here](https://chromium.googlesource.com/chromium/src/tools/gn/+/master/docs/cookbook.md#Common-target-conversion) and [here](https://chromium.googlesource.com/chromium/src/tools/gn/+/master/docs/language.md#Labels). There's also [the comment in .gn itself](https://github.com/crosswalk-project/chromium-crosswalk/blob/crosswalk-22/51.0.2704.105/.gn#L8) explaining what `secondary_source` does.

Quoting from the Language and Operation document:

> When loading a build file, if it doesn‘t exist in the given location relative to the source root, GN will look in the secondary tree in build/secondary. The structure of this tree mirrors the main repository and is a way to add build files for directories that may be pulled from other repositories where we can’t easily check in BUILD files. The secondary tree is a fallback rather than an override, so a file in the normal location always takes precedence.

I don't understand why you are talking about `src/BUILD.gn`. Either `src/xwalk/BUILD.gn` exists and is read from `src/BUILD.gn` or it does not exist (the content shell builds), in which case `src/BUILD.gn` reads `src/build/secondary/xwalk/BUILD.gn`.

With all the above said, another, even simpler approach, is to just change `root_extra_deps` in `src/xwalk/build/xwalk.gni` -- this variable was added to `src/BUILD.gn` precisely for this purpose.

> Another question is, should we enable build content_shell in the environment of both Crosswalk and Chromium-Crosswalk are downloaded?

It must be possible to build content_shell regardless of whether `src/xwalk` is present or not, just like it already happens with gyp.